0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -274,4 +226,72 @@ review_payment(private_headers, psk_data, ilp_packet, ... )
 
 fulfillment_generator = hmac_sha_256(shared_secret, 'ilp_psk_condition')
 fulfillment = hmac_sha_256(fulfillment_generator, ilp_packet)
+
+// The receiver submits the fulfillment to execute the incoming transfer.
+```
+
+## Appendix A: Recommended Algorithm for Generating Shared Secrets
+
+The receiver MAY use any method they choose for determining the `shared_secret`. This algorithm is RECOMMENDED because it allows the receiver to listen for incoming PSK payments with many different shared secrets without needing to store all of the various secret values.
+
+The receiver maintains one `receiver_secret`, which is HMACed with a random public `token` to generate the shared secret initially and regenerate it from incoming payment packets.
+
+### Shared Secret Generation
+
+This is executed by the receiver once per sender, or each time a new `shared_secret` is desired (NOT once per payment):
+
+```js
+/**
+  * @param account (this is the address the receiver will listen on, as
+  * returned by their ledger plugin)
+  */
+
+// The token is attached to the receiver's ILP address, allowing them to
+// regenerate the shared secret every time they receive a payment, instead of
+// keeping a store of all shared secrets they've given out.
+
+receiver_secret = random_bytes(32)
```

Comment:
Maybe `receiver_secret` should be a parameter, because it isn't regenerated every time a shared secret is generated.